### PR TITLE
core: Improve fail fast status messages

### PR DESCRIPTION
**Pull request by ejona86**
Don't wrap the Status when a fail-fast RPC is already queued when the
transport fails, since that Status is directly responsible for the
failing of the RPC. For future fail-fast RPCs, use the saved status as a
cause to make debugging easier.

This came out of #1330, where the unnecessary nesting of the status
codes just added noise.

@buchgr, FYI

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java [hunk 1]
```diff
@@ -79,14 +79,15 @@
...
+   * transport for all addresses, namely the channel is in TRANSIENT_FAILURE. When in a back-off
+   * interval, {@code backoffStatus != null}.
    *
    * <p>If the transport is in a back-off interval, then all fail fast streams (including the
    * pending as well as new ones) will fail immediately. New non-fail fast streams can be created as
    * {@link PendingStream} and will keep pending during this back-off period.
    */
   @GuardedBy("lock")
-  private boolean inBackoffPeriod;
+  private Status backoffStatus;
```
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: @Nullable?

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java [hunk 2]
```diff
@@ -323,8 +324,7 @@ void startBackoff(final Status status) {
           @Override
           public void run() {
             for (PendingStream stream : failFastPendingStreams) {
-              stream.setStream(new FailingClientStream(Status.UNAVAILABLE.withDescription(
-                  "Terminated fail fast stream.").withCause(status.asException())));
+              stream.setStream(new FailingClientStream(status));
```
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: should `new FailingClientStream(backoffStatus)` be better here?

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java [hunk 3]
```diff
@@ -307,8 +307,9 @@ boolean isInBackoffPeriod() {
    */
   void startBackoff(final Status status) {
     synchronized (lock) {
-      Preconditions.checkState(!inBackoffPeriod);
-      inBackoffPeriod = true;
+      Preconditions.checkState(backoffStatus == null);
+      backoffStatus = Status.UNAVAILABLE.withDescription("Channel in TRANSIENT_FAILURE state")
+          .withCause(status.asRuntimeException());
```
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: is this information 
`.withDescription("Channel in TRANSIENT_FAILURE state").withCause(status.asRuntimeException())`
used anywhere?

Comment (ejona86) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: This is passed to RPCs when they fail. So the application will end up seeing this.

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java (hunk 1 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: Done.

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java (hunk 3 above)
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: oh right, just saw it is used in `newStream` method. 
I think we can verify this somewhere in the DelayedTransportTest or
in the testcase 
`TransportSetTest#verifyFailFastAndNonFailFastBehaviors()`
between line# 389-392
```
    // Create a new fail fast stream.
    delayedTransport.newStream(method, headers, failFastCallOptions);  
```
// verify it is a failing stream with back-off information
```
    // Verify it is not queued.
    assertEquals(pendingStreamsCount, delayedTransport.getPendingStreamsCount());
```

Comment (ejona86) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: I went ahead and added a new test to `DelayedClientTransportTest` to check behavior on new streams (and verify the status). I also added to TransportSetTest to have it check one of the statuses to make sure that it was propagated from transportShutdown.

Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: thanks

**Comment by dapengzhang0** [comment]
LGTM

**Pull request merged**